### ISSUE TITLE
🤔 Variadic arguments in New

### ISSUE THREAD

**Issue by peterbourgon** [gofiber/fiber#554]
This is not strictly a bug, but neither is it a feature request; please feel free to reclassify.

```go
// New creates a new Fiber named instance.
// You can pass optional settings when creating a new instance.
func New(settings ...*Settings) *App {
	. . .
```

Variadic arguments `args... T` are essentially a bit of syntax sugar for `args []T`. Their semantics are _zero or more_ of the type. Using them to express _zero or one_ of the type is a misuse of the feature, and confusing to callers. If I read the function signature above, the clear implication of the variadic `settings` argument is that I can pass many of them. But that doesn't make sense in context: which settings would be the one that was used? The first one? The last one? Do you merge them all?

There are a few better and more idiomatic ways to provide the same level of convenience in your exported API.

- Make Settings mandatory, but infer reasonable defaults for zero values: `New(s Settings) *App`
- Two New functions: `New(s Settings) *App` and `NewDefault() *App`
- [Functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis): `New(opt ...AppOption) *App`

**Comment by thomasvvugt** [comment]
Hi @peterbourgon,

I think the current API implementation for the `New` method should remain as-is.
You can pass optional settings when creating a new instance, but maybe we should add more API documentation with examples.

If we really want to get rid of variadic arguments, I would prefer implementing your first suggestion.
As an example:

```go
func New(settings *Settings) *App {
    if settings == nil {
        // Use default settings struct
    } else {
        // Use the provided argument as settings
    }
}
```

Thanks for bringing this to our attention nonetheless!

**Comment by peterbourgon** [comment]
@thomasvvugt 

> I think the current API implementation for the New method should remain as-is. You can pass optional settings when creating a new instance, but maybe we should add more API documentation with examples.

I've provided the rationale for why variadic args are not a good way to expression zero-or-one optionality. The three options I enumerated are, to the best of my knowledge, the only Go-idiomatic ways to express the semantics you're looking for. If that's not convincing to you, then it's not convincing — your project, your decisions :)

> ```go
> func New(settings *Settings) *App {
> ```

I think using `nil` to represent "default settings" is probably no better, and maybe worse, than the current version. `nil` in Go doesn't usually carry the semantics of "not provided" in the way that's implied by this function signature. Callers would invoke it by `app.New(nil)` which is jarring. For one thing, I can't infer the type of that `nil` without going to the docs, which isn't ideal. For another thing, as very few (no?) APIs use `nil` in this way, it scans as a bug.
